app/controllers/core_data_connector/names_controller.rb

```diff
@@ -0,0 +1,43 @@
+module CoreDataConnector
+  class NamesController < ApplicationController
+    # Search attributes
+    search_attributes :name
+
+    def base_query
+      query = super
+
+      return query.none unless params[:project_id]
+
+      nameable_models = [
+        :instance,
+        :item,
+        :work
+      ]
+
+      queries = nameable_models.map do |nm|
+        model_name = "CoreDataConnector::#{nm.to_s.classify}".constantize
+
+        CoreDataConnector::SourceTitle
+          .joins(nm => :project_model)
+          .where(
+            CoreDataConnector::ProjectModel.arel_table[:project_id].eq(params[:project_id])
+            )
+          .where(CoreDataConnector::SourceTitle.arel_table[:nameable_type].eq(model_name))
+          .where(CoreDataConnector::SourceTitle.arel_table[:name_id].eq(CoreDataConnector::Name.arel_table[:id]))
+      end
+
+      raw_sql = <<-SQL.squish
```

Comment:
I like this approach to building the queries since they're so similar. Can we eliminate the need to hard-code array indices? Something like `queries.map(&:to_sql).join(' UNION ')`?

Reply:
👍 Hopefully this was re-tested after refactoring?